nexus/db-model/src/vpc_firewall_rule.rs

```diff
@@ -253,15 +254,42 @@ impl VpcFirewallRule {
     pub fn vec_from_params(
         vpc_id: Uuid,
         params: external::VpcFirewallRuleUpdateParams,
-    ) -> Vec<VpcFirewallRule> {
-        params
+    ) -> Result<Vec<VpcFirewallRule>, external::Error> {
+        ensure_no_duplicates(&params)?;
+        Ok(params
             .rules
             .iter()
             .map(|rule| VpcFirewallRule::new(Uuid::new_v4(), vpc_id, rule))
-            .collect()
+            .collect())
     }
 }
 
+fn ensure_no_duplicates(
+    params: &external::VpcFirewallRuleUpdateParams,
+) -> Result<(), external::Error> {
+    // we could do this by comparing set(names).len() to names.len(), but this
+    // way we can say what the duplicate names are, and that's nice!
+    let mut names = HashSet::new();
+    let mut dupes = HashSet::new();
+    for r in params.rules.iter() {
+        if !names.insert(r.name.clone()) {
+            // insert returns false if already present
+            dupes.insert(r.name.clone());
+        }
+    }
+
+    if dupes.is_empty() {
+        return Ok(());
+    }
+
+    let dupes_str =
+        dupes.iter().map(|d| format!("\"{d}\"")).collect::<Vec<_>>().join(", ");
+    return Err(external::Error::invalid_value(
+        "rules",
+        format!("Rule names must be unique. Duplicates: [{}]", dupes_str),
+    ));
```

Comment:
Honestly, the `fmt::Debug` implementation for `HashSet<String>` is [not _that_ far off from this, except that it'll use `{`/`}` instead of `[`/`]`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2c713dbee1f44a6a50643b43a8fb4c5f), so you probably _could_ just use that, unless you think having square brackets is significantly clearer for the reader. Up to you, though!

Reply:
Cool idea. I tried it and it is tolerable, but it gives me a funny feeling that I traced to not liking that error message output would depend on the choice of iterable data structure.